README.md
# atom-wakatime

[![Coding time tracker](https://wakatime.com/badge/github/wakatime/atom-wakatime.svg)](https://wakatime.com/badge/github/wakatime/atom-wakatime)

Quantify your coding inside GitHub's Atom editor.

## Installation

![install](./install.gif)

1. Inside Atom, navigate to `Preferences`/`Settings` → `Install` and search for `wakatime`.

  ![Project Overview](https://wakatime.com/static/img/ScreenShots/atom-wakatime-install-1.png)

2. Click the `Install` button.

  ![Project Overview](https://wakatime.com/static/img/ScreenShots/atom-wakatime-install-2.png)

3. Click `Settings`, then enter your [api key](https://wakatime.com/settings#apikey).

4. Use Atom and your coding activity will be displayed on your [WakaTime dashboard](https://wakatime.com).

Note: The leet way to install WakaTime is with this Terminal command:

    apm install wakatime


## Screen Shots

![Project Overview](https://wakatime.com/static/img/ScreenShots/Screen-Shot-2016-03-21.png)

## Configuring

Settings for atom-wakatime are under `Settings → Packages → wakatime → Settings`.

Additional settings are in `$HOME/.wakatime.cfg` for [wakatime cli](https://github.com/wakatime/wakatime#configuring).

## Troubleshooting

Atom writes errors to the Atom Developer Console. To check for errors:

1. Turn on the debug checkbox in `Settings → Packages → wakatime → Settings`
2. Inside Atom, go to `View → Developer → Toggle Developer Tools`
3. Clear the developer console (the circle button top left of the console)
4. Inside Atom, go to `View → Developer → Reload Window`

That will reload all plugins including WakaTime, and you should see the startup process logged in the developer console.
If there are no errors in your developer console after startup and editing a file, then check for errors in your `~/.wakatime.log` file.

For checking your `~/.wakatime.log` file, see [wakatime/wakatime#troubleshooting](https://github.com/wakatime/wakatime#troubleshooting).

Note: Behind a proxy? Configure Atom to use your proxy when installing wakatime: <https://github.com/atom/apm#behind-a-firewall>


HISTORY.rst

History
-------


11.0.12 (2023-06-10)
++++++++++++++++++

- Support forks of Atom, for ex: Pulsar.


11.0.11 (2022-04-29)
++++++++++++++++++

- Remove deprecated HTMLDocument.registerElement usage.
  `#123 <https://github.com/wakatime/atom-wakatime/issues/123>`_
- Create symlink to wakatime-cli for consistency with other wakatime plugins.


11.0.10 (2022-03-14)
++++++++++++++++++

- Fix bug where wrong variable name used when creating resources folder.
  `#121 <https://github.com/wakatime/atom-wakatime/issues/121>`_


11.0.9 (2022-01-30)
++++++++++++++++++

- Use separate config file for internal settings.


11.0.8 (2021-12-18)
++++++++++++++++++

- Remove null chars when reading and writing ini cfg file.
  `jetbrains-wakatime#195 <https://github.com/wakatime/jetbrains-wakatime/issues/195>`_


11.0.7 (2021-12-17)
++++++++++++++++++

- Prevent using undefined response in logging.
  `#119 <https://github.com/wakatime/atom-wakatime/issues/119>`_


11.0.6 (2021-11-02)
++++++++++++++++++

- Fix reporting missing wakatime-cli platform support.
- Add platform support for Windows arm64.


11.0.5 (2021-08-14)
++++++++++++++++++

- Add optional chaining operator to http response.


11.0.4 (2021-05-20)
++++++++++++++++++

- Fix caching GitHub releases API requests with correct cfg key name.


11.0.3 (2021-05-20)
++++++++++++++++++

- Prevent deleting wakatime resources folder when downloading new wakatime-cli.
  `#106 <https://github.com/wakatime/atom-wakatime/issues/106>`_


11.0.2 (2021-05-18)
++++++++++++++++++

- Use plugin name in GitHub API User-Agent header, now that ETag not used.


11.0.1 (2021-05-18)
++++++++++++++++++

- GitHub ETag is not reliable, use Last-Modified-Since timestamp instead.


11.0.0 (2021-05-18)
++++++++++++++++++

- Use new Go wakatime-cli.


10.0.0 (2021-01-11)
++++++++++++++++++

- Decaffeinate CoffeeScript into JavaScript.
  `#104 <https://github.com/wakatime/atom-wakatime/issues/104>`_
- Download wakatime-cli async to prevent blocking main thread.
  `#103 <https://github.com/wakatime/atom-wakatime/issues/103>`_


9.0.4 (2021-01-10)
++++++++++++++++++

- Bugfix for Python version checking needs reading both stderr and stdout.
  `#101 <https://github.com/wakatime/atom-wakatime/issues/101>`_


9.0.2 (2020-02-25)
++++++++++++++++++

- Bugfix for getUserHome is not defined error.
  `#98 <https://github.com/wakatime/atom-wakatime/issues/98>`_


9.0.1 (2020-02-25)
++++++++++++++++++

- Respect $WAKATIME_HOME environment variable for config file location.
  `#94 <https://github.com/wakatime/atom-wakatime/issues/94>`_


9.0.0 (2020-02-23)
++++++++++++++++++

- Download wakatime-cli standalone as zipped folder for improved performance.


8.0.6 (2020-02-22)
++++++++++++++++++

- Fix detecting wakatime-cli install filename.


8.0.5 (2020-02-22)
++++++++++++++++++

- Fix attribute used as function.


8.0.4 (2020-02-22)
++++++++++++++++++

- Use process.platform instead of os.platform to detect operating system.
  `#96 <https://github.com/wakatime/atom-wakatime/issues/96>`_


8.0.3 (2020-02-22)
++++++++++++++++++

- Prevent sending heartbeats before wakatime-cli has finished downloading.
  `#96 <https://github.com/wakatime/atom-wakatime/issues/96>`_


8.0.2 (2020-02-22)
++++++++++++++++++

- Add uuid dependency to force latest version and prevent Buffer warning.


8.0.1 (2020-02-22)
++++++++++++++++++

- Fix filename conflict between wakatime-cli and wakatime.coffee.


8.0.0 (2020-02-22)
++++++++++++++++++

- Use standalone wakatime-cli dependency.


7.2.0 (2020-02-09)
++++++++++++++++++

- Detect python in Windows LocalAppData install locations.
- Upgrade embedded python to v3.8.1.


7.1.2 (2019-11-07)
++++++++++++++++++

- Allow python rc versions.
  `#91 <https://github.com/wakatime/atom-wakatime/issues/91>`_
  `#93 <https://github.com/wakatime/atom-wakatime/issues/93>`_


7.1.1 (2019-05-21)
++++++++++++++++++

- Fetch today coding time for status bar when Atom starts.


7.1.0 (2019-05-21)
++++++++++++++++++

- Show today coding time in status bar.


7.0.9 (2019-03-13)
++++++++++++++++++

- Add keywords to package for improved discoverability.


7.0.8 (2019-03-08)
++++++++++++++++++

- Prevent using old Anaconda python distributions because they parse arguments
  containing spaces incorrectly.


7.0.7 (2018-10-03)
++++++++++++++++++

- Improve retry error handling by re-raising original exception.


7.0.6 (2018-10-03)
++++++++++++++++++

- Retry executing wakatime-cli when sending heartbeats up to 5 times.
  `#85 <https://github.com/wakatime/atom-wakatime/issues/85>`_



lib/logger.js
'use babel';

export default class Logger {
  constructor(name) {
    this.levels = {
      ERROR: 40,
      WARN: 30,
      INFO: 20,
      DEBUG: 10,
    };
    this.name = name;
    this.level = 'INFO';
  }

  setLevel(level) {
    try {
      level = level.toUpperCase();
    } catch (error) {}
    if (this.levels[level] == null) {
      const keys = [];
      for (let key in this.levels) {
        keys.push(key);
      }
      this._log('ERROR', `Level must be one of: ${keys.join(', ')}`);
      return;
    }
    this.level = level;
  }

  log(level, msg) {
    level = level.toUpperCase();
    if (this.levels[level] != null && this.levels[level] >= this.levels[this.level]) {
      this._log(level, msg);
    }
  }

  _log(level, msg) {
    level = level.toUpperCase();
    const origLine = this.originalLine();
    if (origLine[0] != null) {
      msg = `[${origLine[0]}:${origLine[1]}] ${msg}`;
    }
    msg = `[${level}] ${msg}`;
    msg = `[${this.name}] ${msg}`;
    switch (level) {
      case 'DEBUG':
        return console.log(msg);
      case 'INFO':
        return console.log(msg);
      case 'WARN':
        return console.warn(msg);
      case 'ERROR':
        return console.error(msg);
    }
  }

  originalLine() {
    const e = new Error('dummy');
    let file = null;
    let line = null;
    let first = true;
    for (let s of e.stack.split('\n')) {
      if (!first) {
        if (s.indexOf('at Logger.') === -1) {
          const m = s.match(/\(?.+[\/\\]([^:]+):(\d+):\d+\)?$/);
          if (m != null) {
            file = m[1];
            line = m[2];
            break;
          }
        }
      }
      first = false;
    }
    return [file, line];
  }

  debug(msg) {
    this.log('DEBUG', msg);
  }

  info(msg) {
    this.log('INFO', msg);
  }

  warn(msg) {
    this.log('WARN', msg);
  }

  error(msg) {
    this.log('ERROR', msg);
  }
}


lib/status-bar-tile-view.js
'use babel';

const element_name = `waka-status-bar${Date.now()}`;

const svg = `\
<svg id="wakatime-svg" width="18px" height="18px" viewBox="0 0 256 256" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" preserveAspectRatio="xMidYMid">
<g><path d="M128,0 C57.308,0 0,57.308 0,128 C0,198.693 57.308,256 128,256 C198.693,256 256,198.693 256,128 C256,57.308 198.693,0 128,0 M128,33.805 C179.939,33.805 222.195,76.061 222.195,128 C222.195,179.94 179.939,222.195 128,222.195 C76.061,222.195 33.805,179.94 33.805,128 C33.805,76.061 76.061,33.805 128,33.805 M115.4993,155.6431 L115.3873,155.6431 C113.4353,155.6081 111.6083,154.6751 110.4343,153.1151 L81.5593,114.7321 C79.4553,111.9351 80.0173,107.9611 82.8143,105.8561 C85.6123,103.7511 89.5853,104.3131 91.6903,107.1111 L115.6833,139.0051 L122.5463,130.5271 C123.7493,129.0411 125.5603,128.1771 127.4723,128.1771 L127.4803,128.1771 C129.3973,128.1791 131.2093,129.0471 132.4103,130.5411 L139.0003,138.7281 L176.6923,90.1341 C178.8353,87.3681 182.8173,86.8651 185.5843,89.0111 C188.3493,91.1561 188.8533,95.1371 186.7073,97.9041 L144.1003,152.8371 C142.9143,154.3681 141.0883,155.2721 139.1503,155.2901 L139.0933,155.2901 C137.1743,155.2901 135.3583,154.4221 134.1553,152.9261 L127.4583,144.6071 L120.4253,153.2931 C119.2213,154.7811 117.4103,155.6431 115.4993,155.6431" fill="#000000"></path></g>
</svg>\
`;
export default class StatusBarTileView {
  constructor () {
    this.element = document.createElement('div');

    this.link = document.createElement('a');
    this.link.classList.add('inline-block');
    this.link.href = 'https://wakatime.com/';
    this.element.appendChild(this.link);

    this.icon = document.createElement('div');
    this.icon.setAttribute('id', 'wakatime-status-bar');
    this.icon.classList.add('inline-block');
    this.icon.innerHTML = svg;
    this.link.appendChild(this.icon);

    this.msg = document.createElement('span');
    this.msg.classList.add('inline-block');
    this.link.appendChild(this.msg);

    this.setStatus('initializing...');
  }

  show() {
    this.element.classList.add(element_name, 'inline-block');
    this.element.classList.remove(element_name, 'hidden');
  }

  hide() {
    this.element.classList.remove(element_name, 'inline-block');
    this.element.classList.add(element_name, 'hidden');
  }

  destroy() {
    if (this.tooltip != null) {
      this.tooltip.dispose();
    }
    this.element.classList = '';
  }

  setStatus(content, override_errors) {
    if (override_errors == null) {
      override_errors = true;
    }
    if (
      override_errors ||
      /^error/i.test((this.msg != null ? this.msg.textContent : undefined) || '')
    ) {
      this.msg != null ? (this.msg.textContent = content || '') : undefined;
    }
  }

  setTitle(text) {
    if (this.tooltip != null) {
      this.tooltip.dispose();
    }
    this.tooltip = atom.tooltips.add(this.element, { title: text });
  }
}

lib/wakatime.js
'use babel';

/*
WakaTime
Description: Analytics for programmers.
Maintainer:  WakaTime <support@wakatime.com>
License:     BSD, see LICENSE for more details.
Website:     https://wakatime.com/
*/
import decompress from 'decompress';
import del from 'del';
import pathExists from 'path-exists';
import fs from 'fs';
import os from 'os';
import path from 'path';
import process from 'process';
import child_process from 'child_process';
import request from 'request';

import StatusBarTileView from './status-bar-tile-view';
import Logger from './logger';
import Options from './options';

// package-global attributes
let log = null;
let options = null;
let packageName = null;
let packageVersion = null;
let lastHeartbeat = 0;
let lastFile = '';
let lastTodayFetch = 0;
const fetchTodayInterval = 60000;
let cachedToday = '';
let statusBarIcon = null;
let pluginReady = false;
let settingChangedObserver = null;
let statusBarTile = null;
let resourcesLocation = null;
let latestCliVersion = '';
const githubDownloadPrefix = 'https://github.com/wakatime/wakatime-cli/releases/download';
const githubReleasesStableUrl = 'https://api.github.com/repos/wakatime/wakatime-cli/releases/latest';
const githubReleasesAlphaUrl = 'https://api.github.com/repos/wakatime/wakatime-cli/releases?per_page=1';

export function activate() {
  log = new Logger('WakaTime');
  options = new Options(getHomeDirectory());
  if (atom.config.get('wakatime.debug')) {
    log.setLevel('DEBUG');
  }
  packageName = atom.getAppName().toLocaleLowerCase().replace(' ', '_');
  packageVersion = atom.packages.getLoadedPackage('wakatime').metadata.version;
  log.debug(`Initializing WakaTime v${packageVersion}...`);
  requestIdleCallback(delayedActivate, { timeout: 10000 });
}

function delayedActivate() {
  setupConfigs();
  settingChangedObserver = atom.config.observe('wakatime', settingChangedHandler);
  checkCLI();
}

export function consumeStatusBar(statusBar) {
  statusBarIcon = new StatusBarTileView();
  statusBarTile =
    statusBar != null ? statusBar.addRightTile({ item: statusBarIcon, priority: 300 }) : undefined;

  // set status bar icon visibility
  if (atom.config.get('wakatime.showStatusBarIcon')) {
    statusBarIcon.show();
  } else {
    statusBarIcon.hide();
  }

  if (pluginReady) {
    statusBarIcon.setTitle('WakaTime ready');
    statusBarIcon.setStatus(cachedToday);
  }
}

export function deactivate() {
  if (statusBarTile != null) {
    statusBarTile.destroy();
  }
  if (statusBarIcon != null) {
    statusBarIcon.destroy();
  }
  settingChangedObserver != null ? settingChangedObserver.dispose() : undefined;
}

function checkCLI() {
  if (!isCLIInstalled()) {
    installCLI(() => {
      log.debug('Finished installing wakatime-cli.');
      finishActivation();
    });
  } else {
    getLastCheckedForUpdates((lastChecked) => {
      // only check for updates to wakatime-cli every 24 hours
      const hours = 24;

      const currentTime = Math.round(new Date().getTime() / 1000);
      const beenLongEnough = lastChecked + 3600 * hours < currentTime;

      if (beenLongEnough || atom.config.get('wakatime.debug')) {
        setLastCheckedForUpdates(currentTime);
        isCLILatest((latest) => {
          if (!latest) {
            installCLI(() => {
              log.debug('Finished installing wakatime-cli.');
              finishActivation();
            });
          } else {
            finishActivation();
          }
        });
      } else {
        finishActivation();
      }
    });
  }
}

function getLastCheckedForUpdates(callback) {
  const filePath = path.join(extensionFolder(), 'last-checked-for-updates');
  if (fs.existsSync(filePath)) {
    return fs.readFile(filePath, 'utf-8', (err, contents) => {
      if (err != null) {
        if (callback != null) {
          callback(0);
        }
        return;
      }
      if (contents != null) {
        try {
          if (callback != null) {
            callback(parseInt(contents.trim(), 10) || 0);
          }
          return;
        } catch (error) {}
      }
      if (callback != null) {
        return callback(0);
      }
    });
  } else {
    if (callback != null) {
      return callback(0);
    }
  }
}

function setLastCheckedForUpdates(lastChecked) {
  const filePath = path.join(extensionFolder(), 'last-checked-for-updates');
  fs.writeFile(filePath, lastChecked.toString(), { encoding: 'utf-8' }, (err) => {
    if (err != null) {
      log.debug('Unable to save last checked for updates timestamp.');
    }
  });
}

function finishActivation() {
  pluginReady = true;
  setupEventHandlers();

  // set status bar icon visibility
  if (atom.config.get('wakatime.showStatusBarIcon')) {
    if (statusBarIcon != null) {
      statusBarIcon.show();
    }
  } else {
    if (statusBarIcon != null) {
      statusBarIcon.hide();
    }
  }

  if (statusBarIcon != null) {
    statusBarIcon.setTitle('WakaTime ready');
  }
  if (statusBarIcon != null) {
    statusBarIcon.setStatus(cachedToday);
  }
  getToday();
  log.debug('Finished initializing WakaTime.');
}

function settingChangedHandler(settings, initial) {
  if (settings.showStatusBarIcon) {
    if (statusBarIcon != null) {
      statusBarIcon.show();
    }
  } else {
    if (statusBarIcon != null) {
      statusBarIcon.hide();
    }
  }
  if (atom.config.get('wakatime.debug')) {
    log.setLevel('DEBUG');
  } else {
    log.setLevel('INFO');
  }
  const apiKey = settings.apikey;
  if (isValidApiKey(apiKey)) {
    atom.config.set('wakatime.apikey', ''); // clear setting so it updates in UI
    atom.config.set('wakatime.apikey', 'Saved in your $WAKATIME_HOME/.wakatime.cfg file');
    options.setSetting('settings', 'api_key', apiKey, false);
  } else if (initial) {
    atom.config.set('wakatime.apikey', ''); // clear setting so it updates in UI
    atom.config.set('wakatime.apikey', 'Enter your api key...');
  }
}

function setupConfigs() {
  options.getSetting('settings', 'api_key', false, (key) => {
    if (isValidApiKey(key.value)) {
      atom.config.set('wakatime.apikey', ''); // clear setting so it updates in UI
      atom.config.set('wakatime.apikey', 'Saved in your $WAKATIME_HOME/.wakatime.cfg file');
    } else {
      settingChangedHandler(atom.config.get('wakatime'), true);
    }
  });
}

function isValidApiKey(key) {
  if (key == null) {
    return false;
  }
  const re = new RegExp(
    '^(waka_)?[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$',
    'i',
  );
  return re.test(key);
}

const enoughTimePassed = (time) => lastHeartbeat + 120000 < time;

function setupEventHandlers(callback) {
  return atom.workspace.observeTextEditors((editor) => {
    try {
      const buffer = editor.getBuffer();
      buffer.onDidSave((e) => {
        const { file } = buffer;
        if (file != null && file) {
          let lineno = null;
          if (editor.cursors.length > 0) {
            lineno = editor.cursors[0].getCurrentLineBufferRange().end.row + 1;
          }
          return sendHeartbeat(file, lineno, true);
        }
      });
      buffer.onDidChange((e) => {
        const { file } = buffer;
        if (file != null && file) {
          let lineno = null;
          if (editor.cursors.length > 0) {
            lineno = editor.cursors[0].getCurrentLineBufferRange().end.row + 1;
          }
          return sendHeartbeat(file, lineno);
        }
      });
      editor.onDidChangeCursorPosition((e) => {
        const { file } = buffer;
        if (file != null && file) {
          let lineno = null;
          if (editor.cursors.length > 0) {
            lineno = editor.cursors[0].getCurrentLineBufferRange().end.row + 1;
          }
          return sendHeartbeat(file, lineno);
        }
      });
    } catch (error) {}
    if (callback != null) {
      return callback();
    }
  });
}

function isCLIInstalled() {
  return fs.existsSync(getCliLocation());
}

function isCLILatest(callback) {
  const args = ['--version'];
  child_process.execFile(getCliLocation(), args, (error, stdout, stderr) => {
    if (error == null) {
      const currentVersion = stdout.trim() + stderr.trim();
      log.debug(`Current wakatime-cli version is ${currentVersion}`);
      log.debug('Checking for updates to wakatime-cli...');
      getLatestCliVersion((latestVersion) => {
        if (currentVersion === latestVersion) {
          log.debug('wakatime-cli is up to date.');
          if (callback) callback(true);
        } else {
          if (latestVersion != null) {
            log.debug(`Found an updated wakatime-cli ${latestVersion}`);
            if (callback) callback(false);
          } else {
            log.debug('Unable to find latest wakatime-cli version.');
            if (callback) callback(true);
          }
        }
      });
    } else {
      if (callback) callback(false);
    }
  });
}

function getLatestCliVersion(callback) {
  if (latestCliVersion) {
    callback(latestCliVersion);
    return;
  }

  options.getSetting('settings', 'proxy', false, (proxy) => {
    options.getSetting('settings', 'no_ssl_verify', false, (noSSLVerify) => {
      options.getSetting('internal', 'cli_version_last_modified', true, (modified) => {
        options.getSetting('settings', 'alpha', false, (alpha) => {
          const opt = {
            url: alpha.value == 'true' ? githubReleasesAlphaUrl : githubReleasesStableUrl,
            json: true,
            headers: {
              'User-Agent': 'github.com/wakatime/atom-wakatime',
            },
          };
          if (proxy.value) opt['proxy'] = proxy.value;
          if (!atom.config.get('wakatime.disableSSLCertVerify') || noSSLVerify.value === 'true') opt['strictSSL'] = false;
          if (modified.value) opt['headers']['If-Modified-Since'] = modified.value;
          try {
            request.get(opt, (error, response, json) => {
              if (!error && response && (response.statusCode == 200 || response.statusCode == 304)) {
                log.debug(`GitHub API Response ${response.statusCode}`);
                if (response.statusCode == 304) {
                  options.getSetting('internal', 'cli_version', true, (version) => {
                    latestCliVersion = version.value;
                    callback(latestCliVersion);
                  });
                  return;
                }
                latestCliVersion = alpha.value == 'true' ? json[0]['tag_name'] : json['tag_name'];
                log.debug(`Latest wakatime-cli version from GitHub: ${latestCliVersion}`);
                const lastModified = response.headers['last-modified'];
                if (lastModified) {
                  options.setSettings('internal', true, [
                    {key: 'cli_version', value: latestCliVersion},
                    {key: 'cli_version_last_modified', value: lastModified},
                  ]);
                }
                callback(latestCliVersion);
                return;
              } else {
                if (response) {
                  log.warn(`GitHub API Response ${"statusCode" in response ? response.statusCode : ""}: ${error}`);
                } else {
                  log.warn(`GitHub API Request Error: ${error}`);
                }
                callback('');
              }
            });
          } catch (e) {
            log.warn(e);
            callback('');
          }
        });
      });
    });
  });
}

function getResourcesLocation() {
  if (resourcesLocation) return resourcesLocation;

  resourcesLocation = path.join(getHomeDirectory(), '.wakatime');
  try {
    fs.mkdirSync(resourcesLocation, { recursive: true });
  } catch (e) {
    log.error(e);
  }
  return resourcesLocation;
}

function getConfigFile(internal) {
  if (internal) return path.join(getHomeDirectory(), '.wakatime-internal.cfg');
  return path.join(getHomeDirectory(), '.wakatime.cfg');
}

function getHomeDirectory() {
  let home = process.env.WAKATIME_HOME;
  if (home && home.trim() && fs.existsSync(home.trim())) return home.trim();
  return process.env[isWindows() ? 'USERPROFILE' : 'HOME'] || '';
}

function extensionFolder() {
  return __dirname;
}

function getCliLocation() {
  const ext = isWindows() ? '.exe' : '';
  const osname = getOS();
  const arch = architecture();
  return path.join(getResourcesLocation(), `wakatime-cli-${osname}-${arch}${ext}`);
}

function isSymlink(file) {
  try {
    return fs.lstatSync(file).isSymbolicLink();
  } catch (_) {}
  return false;
}

function installCLI(callback) {
  getLatestCliVersion(version => {
    const url = cliDownloadUrl(version);
    log.debug(`Downloading wakatime-cli from ${url}`);
    if (statusBarIcon != null) {
      statusBarIcon.setStatus('downloading wakatime-cli...');
    }
    const zipFile = path.join(getResourcesLocation(), 'wakatime-cli.zip');
    downloadFile(url, zipFile, async () => {
      await extractCLI(zipFile);
      const cli = getCliLocation();
      try {
        log.debug('Chmod 755 wakatime-cli...');
        fs.chmodSync(cli, 0o755);
      } catch (e) {
        log.warn(e);
      }
      const ext = isWindows() ? '.exe' : '';
      const link = path.join(getResourcesLocation(), `wakatime-cli${ext}`);
      if (!isSymlink(link)) {
        try {
          log.debug(`Create symlink from wakatime-cli to ${cli}`);
          fs.symlinkSync(cli, link);
        } catch (e) {
          log.warn(e);
          try {
            fs.copyFileSync(cli, link);
            fs.chmodSync(link, 0o755);
          } catch (e2) {
            log.warn(e2);
          }
        }
      }
    });
  });
}

async function extractCLI(zipFile) {
  log.debug('Extracting wakatime-cli.zip file...');
  if (statusBarIcon != null) {
    statusBarIcon.setStatus('extracting wakatime-cli...');
  }
  await removeCLI();
  await unzip(zipFile, getResourcesLocation());
}

async function removeCLI() {
  try {
    await del([getCliLocation()], { force: true });
  } catch (e) {
    log.warn(e);
  }
}

async function unzip(file, outputDir) {
  if (await pathExists(file)) {
    try {
      await decompress(file, outputDir);
    } catch (e) {
      log.warn(e);
    } finally {
      try {
        await del([file], { force: true });
      } catch (err) {
        log.warn(err);
      }
    }
  }
}

function isWindows() {
  return os.platform() === 'win32';
}

function getOS() {
  let name = `${os.platform()}`;
  if (name == 'win32') name = 'windows';
  return name;
}

function architecture() {
  const arch = os.arch();
  if (arch.indexOf('arm') > -1) return arch;
  if (arch.indexOf('32') > -1) return '386';
  return 'amd64';
}

function cliDownloadUrl(version) {
  const osname = getOS();
  const arch = architecture();

  const validCombinations = [
    'darwin-amd64',
    'darwin-arm64',
    'freebsd-386',
    'freebsd-amd64',
    'freebsd-arm',
    'linux-386',
    'linux-amd64',
    'linux-arm',
    'linux-arm64',
    'netbsd-386',
    'netbsd-amd64',
    'netbsd-arm',
    'openbsd-386',
    'openbsd-amd64',
    'openbsd-arm',
    'openbsd-arm64',
    'windows-386',
    'windows-amd64',
    'windows-arm64',
  ];
  if (!validCombinations.includes(`${osname}-${arch}`)) reportMissingPlatformSupport(osname, arch);

  return `${githubDownloadPrefix}/${version}/wakatime-cli-${osname}-${arch}.zip`;
}

function reportMissingPlatformSupport(osname, architecture) {
  const url = `https://api.wakatime.com/api/v1/cli-missing?osnameame=${osname}&architecture=${architecture}&plugin=${packageName}`;
  let opt = {
    url: url,
    strictSSL: !atom.config.get('wakatime.disableSSLCertVerify'),
    ca: fs.readFileSync(path.join(extensionFolder(), 'cafile.pem')),
  };
  options.getSetting('settings', 'proxy', false, (proxy) => {
    if (proxy.value) options['proxy'] = proxy.value;
    try {
      request.get(opt);
    } catch (e) { }
  });
}

function downloadFile(url, outputFile, callback) {
  const opt = {
    strictSSL: !atom.config.get('wakatime.disableSSLCertVerify'),
    url,
  };
  options.getSetting('settings', 'proxy', false, (proxy) => {
    if (proxy.value) options['proxy'] = proxy.value;
    const r = request(opt);
    const out = fs.createWriteStream(outputFile);
    r.pipe(out);
    r.on('end', () =>
      out.on('finish', async () => {
        if (callback != null) {
          return await callback();
        }
      }),
    );
  });
}

function sendHeartbeat(file, lineno, isWrite) {
  if (!isCLIInstalled()) {
    return;
  }

  if (
    (file.path == null || file.path === undefined) &&
    (file.getPath == null || file.getPath() === undefined)
  ) {
    log.debug(`Skipping file because path does not exist: ${file.path}`);
    return;
  }

  const currentFile = file.path || file.getPath();

  if (fileIsIgnored(currentFile)) {
    log.debug(`Skipping file because path matches ignore pattern: ${currentFile}`);
    return;
  }

  const time = Date.now();
  if (isWrite || enoughTimePassed(time) || lastFile !== currentFile) {
    const args = ['--entity', currentFile, '--plugin', `${packageName}-wakatime/${packageVersion}`];
    if (isWrite) {
      args.push('--write');
    }
    if (lineno != null) {
      args.push('--lineno');
      args.push(lineno);
    }
    if (atom.config.get('wakatime.debug')) {
      args.push('--verbose');
    }
    if (atom.config.get('wakatime.disableSSLCertVerify')) {
      args.push('--no-ssl-verify');
    }

    // fix for wakatime/atom-wakatime#65
    args.push('--config');
    args.push(getConfigFile());

    if (atom.project.contains(currentFile)) {
      for (let rootDir of atom.project.rootDirectories) {
        const { realPath } = rootDir;
        if (currentFile.indexOf(realPath) > -1) {
          args.push('--alternate-project');
          args.push(path.basename(realPath));
          break;
        }
      }
    }

    lastHeartbeat = time;
    lastFile = currentFile;

    const cli = getCliLocation();
    log.debug(`${cli} ${args.join(' ')}`);
    executeHeartbeatProcess(cli, args, 0);
    return getToday();
  }
}

function executeHeartbeatProcess(binary, args, tries) {
  const max_retries = 5;
  try {
    const proc = child_process.execFile(binary, args, (error, stdout, stderr) => {
      if (error != null) {
        let msg, status, title;
        if (stderr != null && stderr !== '') {
          log.warn(stderr);
        }
        if (stdout != null && stdout !== '') {
          log.warn(stdout);
        }
        if (proc.exitCode === 102 || proc.exitCode === 112) {
          msg = null;
          status = null;
          title = 'WakaTime Offline, coding activity will sync when online.';
        } else if (proc.exitCode === 103) {
          msg =
            'An error occured while parsing $WAKATIME_HOME/.wakatime.cfg. Check $WAKATIME_HOME/.wakatime.log for more info.';
          status = 'Error';
          title = msg;
        } else if (proc.exitCode === 104) {
          msg = 'Invalid API Key. Make sure your API Key is correct!';
          status = 'Error';
          title = msg;
        } else {
          msg = error;
          status = 'Error';
          title = `Unknown Error (${proc.exitCode}); Check your Dev Console and ~/.wakatime.log for more info.`;
        }

        if (msg != null) {
          log.warn(msg);
        }
        if (statusBarIcon != null) {
          statusBarIcon.setStatus(status);
        }
        return statusBarIcon != null ? statusBarIcon.setTitle(title) : undefined;
      } else {
        if (statusBarIcon != null) {
          statusBarIcon.setStatus(cachedToday);
        }
        const today = new Date();
        return statusBarIcon != null
          ? statusBarIcon.setTitle(`Last heartbeat sent ${formatDate(today)}`)
          : undefined;
      }
    });
  } catch (e) {
    tries++;
    const retry_in = 2;
    if (tries < max_retries) {
      log.debug(
        `Failed to send heartbeat when executing wakatime-cli background process, will retry in ${retry_in} seconds...`,
      );
      setTimeout(() => executeHeartbeatProcess(binary, args, tries), retry_in * 1000);
    } else {
      log.error('Failed to send heartbeat when executing wakatime-cli background process.');
      throw e;
    }
  }
}

function getToday() {
  if (!isCLIInstalled()) {
    return;
  }

  const cutoff = Date.now() - fetchTodayInterval;
  if (lastTodayFetch > cutoff) {
    return;
  }
  lastTodayFetch = Date.now();

  const args = ['--today', '--plugin', `${packageName}-wakatime/${packageVersion}`];
  if (atom.config.get('wakatime.disableSSLCertVerify')) {
    args.push('--no-ssl-verify');
  }
  args.push('--config');
  args.push(getConfigFile());

  try {
    child_process.execFile(getCliLocation(), args, (error, stdout, stderr) => {
      if (error != null) {
        if (stderr != null && stderr !== '') {
          log.debug(stderr);
        }
        if (stdout != null && stdout !== '') {
          log.debug(stderr);
        }
      } else {
        cachedToday = `Today: ${stdout}`;
        statusBarIcon != null ? statusBarIcon.setStatus(cachedToday, true) : undefined;
      }
    });
  } catch (e) {
    log.debug(e);
  }
}

function fileIsIgnored(file) {
  if (
    endsWith(file, 'COMMIT_EDITMSG') ||
    endsWith(file, 'PULLREQ_EDITMSG') ||
    endsWith(file, 'MERGE_MSG') ||
    endsWith(file, 'TAG_EDITMSG')
  ) {
    return true;
  }
  const patterns = atom.config.get('wakatime.ignore');
  if (patterns == null) {
    return true;
  }

  let ignore = false;
  for (let pattern of patterns) {
    const re = new RegExp(pattern, 'gi');
    if (re.test(file)) {
      ignore = true;
      break;
    }
  }
  return ignore;
}

function endsWith(str, suffix) {
  if (str != null && suffix != null) {
    return str.indexOf(suffix, str.length - suffix.length) !== -1;
  }
  return false;
}

function formatDate(date) {
  const months = [
    'Jan',
    'Feb',
    'Mar',
    'Apr',
    'May',
    'Jun',
    'Jul',
    'Aug',
    'Sep',
    'Oct',
    'Nov',
    'Dec',
  ];
  let ampm = 'AM';
  let hour = date.getHours();
  if (hour > 11) {
    ampm = 'PM';
    hour = hour - 12;
  }
  if (hour === 0) {
    hour = 12;
  }
  let minute = date.getMinutes();
  if (minute < 10) {
    minute = `0${minute}`;
  }
  return `${
    months[date.getMonth()]
  } ${date.getDate()}, ${date.getFullYear()} ${hour}:${minute} ${ampm}`;
}
